### PR TITLE
Remove not needed database updates in modify user admin API

### PR DESCRIPTION
**Pull request by dklimpel**
Follow up: #10598 - https://github.com/matrix-org/synapse/pull/10598#discussion_r688582506

Update only changed threepids and not all.
Add type hints, update test.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: synapse/rest/admin/users.py [hunk 1]
```diff
@@ -229,37 +229,50 @@ async def on_PUT(
...
             for external_id in external_ids:
                 new_external_ids.append(
                     (external_id["auth_provider"], external_id["external_id"])
                 )
 
+        # convert List[Dict[str, str]] into List[Tuple[str, str]]
+        if threepids is not None:
+            new_threepids = []
+            for threepid in threepids:
+                new_threepids.append((threepid["medium"], threepid["address"]))
```
Comment (richvdh) [inline synapse/rest/admin/users.py:244]: why not construct these as sets?

```suggestion
        # convert List[Dict[str, str]] into Set[Tuple[str, str]]
        if threepids is not None:
            new_threepids = {
                (threepid["medium"], threepid["address"])
                for threepid in threepids
            }
```

Likewise `cur_threepids` below - then you can remove the `set()` operations.

File: synapse/storage/databases/main/registration.py [hunk 2]
```diff
@@ -754,16 +754,18 @@ async def get_user_id_by_threepid(self, medium: str, address: str) -> Optional[s
...
         Args:
             txn (cursor):
-            medium (str): threepid medium e.g. email
-            address (str): threepid address e.g. me@example.com
+            medium: threepid medium e.g. email
+            address: threepid address e.g. me@example.com
 
         Returns:
-            str|None: user id or None if no user id/threepid mapping exists
+            user id or None if no user id/threepid mapping exists
```
Comment (richvdh) [inline synapse/storage/databases/main/registration.py:768]:
```suggestion
            user id, or None if no user id/threepid mapping exists
```

**Review by richvdh — COMMENTED** [review]
looks good, though a stylistic suggestion here.

File: synapse/rest/admin/users.py (hunk 1 above)
Comment (dklimpel) [reply synapse/rest/admin/users.py]: I just didn't think far enough. :(

**Review by richvdh — APPROVED** [review]
lgtm, thank you!

**Pull request merged**